File: Backend/src/main/java/backend/osiris/entity/inventory/PurchaseOrder.java
package backend.osiris.entity.inventory;

import backend.osiris.entity.BaseEntity;
import backend.osiris.entity.product.Supplier;
import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@Accessors(chain = true)
@Entity
@Table(name = "purchase_order")
public class PurchaseOrder extends BaseEntity {
    @Column(name = "code", nullable = false, unique = true)
    private String code;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "supplier_id", nullable = false)
    @JsonBackReference
    private Supplier supplier;

    @OneToMany(mappedBy = "purchaseOrder", cascade = CascadeType.ALL)
    private Set<PurchaseOrderVariant> purchaseOrderVariants = new HashSet<>();

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "destination_id", nullable = false)
    @JsonBackReference
    private Destination destination;

    @Column(name = "total_amount", nullable = false)
    private Double totalAmount;

    @Column(name = "note")
    private String note;

    @Column(name = "status", nullable = false, columnDefinition = "TINYINT")
    private Integer status;

    @OneToMany(mappedBy = "purchaseOrder", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonManagedReference
    private List<Docket> dockets = new ArrayList<>();
}

File: Frontend/src/models/PurchaseOrder.ts
import BaseResponse from "./BaseResponse";
import { DestinationResponse } from "./Destination";
import {
    PurchaseOrderVariantResponse,
    PurchaseOrderVariantRequest,
} from "./PurchaseOrderVariant";
import { SupplierResponse } from "./Supplier";

export interface PurchaseOrderResponse extends BaseResponse {
    createdBy: any;
    updatedBy: any;
    code: string;
    supplier: SupplierResponse;
    purchaseOrderVariants: PurchaseOrderVariantResponse[];
    destination: DestinationResponse;
    totalAmount: number;
    note: string | null;
    status: number;
    dockets: DocketResponse[];
}

interface DocketResponse extends BaseResponse {
    type: number;
    code: string;
    warehouse: WarehouseResponse;
    status: number;
}

interface WarehouseResponse extends BaseResponse {
    code: string;
    name: string;
    status: number;
}

export interface PurchaseOrderRequest {
    code: string;
    supplierId: number;
    purchaseOrderVariants: PurchaseOrderVariantRequest[];
    destinationId: number;
    totalAmount: number;
    note: string | null;
    status: number;
}
